README.md
# NAME

Mew - Moo with sugar on top

# SYNOPSIS

<div>
    <div style="display: table; height: 91px; background: url(http://zoffix.com/CPAN/Dist-Zilla-Plugin-Pod-Spiffy/icons/section-code.png) no-repeat left; padding-left: 120px;" ><div style="display: table-cell; vertical-align: middle;">
</div>

    # This:
    use Mew;
    has  _foo  => PositiveNum;
    has -_bar  => Bool;  # note the minus: it means attribute is not `required`
    has  _type => ( Str, default => 'text/html', chained => 1);
    has  _cust => ( is => 'ro', isa => sub{ 42 } ); # standard Moo `has`

    # Is same as:
    use strictures 2;
    use Types::Standard qw/:all/;
    use Types::Common::Numeric qw/:all/;
    use Moo;
    use MooX::ChainedAttributes;
    use namespace::clean;

    has _foo  => (
        init_arg => 'foo',
        is       => 'ro'
        isa      => PositiveNum,
        required => 1,
    );
    has _bar  => (
        init_arg => 'bar',
        is       => 'ro'
        isa      => Bool,
    );
    has _type  => (
        chained  => 1,
        init_arg => 'type',
        is       => 'rw'
        isa      => Str,
        default  => 'text/html',
    );
    has _cust => (
        is  => 'ro',
        isa => sub{ 42 },
    );

<div>
    </div></div>
</div>

# DESCRIPTION

This module is just like regular [Moo](https://metacpan.org/pod/Moo), except it also imports
[strictures](https://metacpan.org/pod/strictures) and [namespace::clean](https://metacpan.org/pod/namespace::clean), along with
a couple of standard types modules. In addition, it sweetens the
[Moo's has subroutine](https://metacpan.org/pod/Moo#has) to allow for more concise attribute
declarations.

# READ FIRST

Virtually all of the functionality is described in [Moo](https://metacpan.org/pod/Moo).

# IMPORTED MODULES

    use Mew;

Automatically imports the following modules: [Moo](https://metacpan.org/pod/Moo), [strictures](https://metacpan.org/pod/strictures),
[Types::Standard](https://metacpan.org/pod/Types::Standard), [Types::Common::Numeric](https://metacpan.org/pod/Types::Common::Numeric), [MooX::ChainedAttributes](https://metacpan.org/pod/MooX::ChainedAttributes),
and [namespace::clean](https://metacpan.org/pod/namespace::clean). **NOTE: in particular the last one.** It'll scrub
your namespace, thus if you're using things like [experimental](https://metacpan.org/pod/experimental), you should
declare them **after** you `use Mew`.

# `has` SUGAR

## Call it like if it were Moo

    has _cust => ( is => 'ro' );

First, you can call `has` just like you'd call ["has" in Moo](https://metacpan.org/pod/Moo#has) and it'll work
exactly as it used to. The sugar won't be enabled in that case.

## Specify `isa` type to get sugar

    has _cust => Str;
    has _cust => Str, ( default => "foo" ); # Note: can't use "=>" after Str

To get the sugar, you need to specify one of the imported types from either
[Types::Standard](https://metacpan.org/pod/Types::Standard) or [Types::Common::Numeric](https://metacpan.org/pod/Types::Common::Numeric) as the second argument. Once
that is done, `Mew` will add some default settings, which are:

    1) Set `isa` to the type you gave
    2) Set `is` to 'ro' (or 'rw', if `chained` is set)
    3) Set `require` to 1
    4) Set `init_arg` to the name of the attribute, removing
        the leading underscore, if it's present

Thus, `has _cust => Str;` is equivalent to

    use Types::Standard qw/Str/;
    has _cust => (
        init_arg => 'cust',
        is       => 'ro'
        isa      => Str,
        required => 1,
    );

<div>
    <div style="display: table; height: 91px; background: url(http://zoffix.com/CPAN/Dist-Zilla-Plugin-Pod-Spiffy/icons/section-warning.png) no-repeat left; padding-left: 120px;" ><div style="display: table-cell; vertical-align: middle;">
</div>

**IMPORTANT NOTE:** because Perl's fat comma (`=>`) quotes the argument
on the left side, using it after the type won't work:

    has _cust => Str => ( default => "BROKEN" ); # WRONG!!!!
    has _cust => Str, ( default => "WORKS" ); # Correct!
    has _cust => ( Str, default => "WORKS" ); # This is fine too

<div>
    </div></div>
</div>

## Method chaining

    package Foo;
    use Mew;
    has cost   => ( PostiveNum, chained => 1 );
    has weight => ( PostiveNum, chained => 1 );
    has size   => ( Str,        chained => 1 );

    ...

    my $object = Foo->new->cost( 42 )->weight( 45 )->size("X-Large");
    say $object->size; # prints "X-Large"

To have [fluent interface](https://en.wikipedia.org/wiki/Fluent_interface)
or allow "chaining" your attributes, simply add `chained => 1` option
to your attribute declaration. **Note:** this will automatically use
`rw` instead of `ro` for the default of the `is` option.

### Modify the sugar

It's possible to alter the defaults created by `Mew`:

#### Remove `required`

    has -_cust => Str;

Simply prefix the attribute's name with a minus sign to avoid setting
`required => 1`.

#### Modify other options

    has  _cust => Str, ( init arg => "bar" );
    has -_cust => Str, ( is => "lazy" );

You can explicitly provide values for options set by `Mew`, in which case
the values you provide will be used instead of the defaults.

# SEE ALSO

[Moo](https://metacpan.org/pod/Moo), [Type::Tiny](https://metacpan.org/pod/Type::Tiny)

<div>
    <div style="background: url(http://zoffix.com/CPAN/Dist-Zilla-Plugin-Pod-Spiffy/icons/hr.png);height: 18px;"></div>
</div>

# REPOSITORY

<div>
    <div style="display: table; height: 91px; background: url(http://zoffix.com/CPAN/Dist-Zilla-Plugin-Pod-Spiffy/icons/section-github.png) no-repeat left; padding-left: 120px;" ><div style="display: table-cell; vertical-align: middle;">
</div>

Fork this module on GitHub:
[https://github.com/zoffixznet/Mew](https://github.com/zoffixznet/Mew)

<div>
    </div></div>
</div>

# BUGS

<div>
    <div style="display: table; height: 91px; background: url(http://zoffix.com/CPAN/Dist-Zilla-Plugin-Pod-Spiffy/icons/section-bugs.png) no-repeat left; padding-left: 120px;" ><div style="display: table-cell; vertical-align: middle;">
</div>

To report bugs or request features, please use
[https://github.com/zoffixznet/Mew/issues](https://github.com/zoffixznet/Mew/issues)

If you can't access GitHub, you can email your request
to `bug-Mew at rt.cpan.org`

<div>
    </div></div>
</div>

# AUTHOR

Part of the code was borrowed from [Moo](https://metacpan.org/pod/Moo)'s innards. [ew](https://metacpan.org/pod/ew) module is an
almost-verbatim copy of [oo](https://metacpan.org/pod/oo) module. Thanks to _Matt S. Trout (mst)_ for
changing my copypasta of Moo's internals to sane code and other help.
Props to _Altreus_ for coming up with the name for the module.

The rest is:

<div>
    <div style="display: table; height: 91px; background: url(http://zoffix.com/CPAN/Dist-Zilla-Plugin-Pod-Spiffy/icons/section-author.png) no-repeat left; padding-left: 120px;" ><div style="display: table-cell; vertical-align: middle;">
</div>

<div>
    <span style="display: inline-block; text-align: center;"> <a href="http://metacpan.org/author/ZOFFIX"> <img src="http://www.gravatar.com/avatar/328e658ab6b08dfb5c106266a4a5d065?d=http%3A%2F%2Fwww.gravatar.com%2Favatar%2F627d83ef9879f31bdabf448e666a32d5" alt="ZOFFIX" style="display: block; margin: 0 3px 5px 0!important; border: 1px solid #666; border-radius: 3px; "> <span style="color: #333; font-weight: bold;">ZOFFIX</span> </a> </span>
</div>

<div>
    </div></div>
</div>

# CONTRIBUTORS

<div>
    <div style="display: table; height: 91px; background: url(http://zoffix.com/CPAN/Dist-Zilla-Plugin-Pod-Spiffy/icons/section-contributors.png) no-repeat left; padding-left: 120px;" ><div style="display: table-cell; vertical-align: middle;">
</div>

<div>
    <span style="display: inline-block; text-align: center;"> <a href="http://metacpan.org/author/MSTROUT"> <img src="http://www.gravatar.com/avatar/9a085716bde55f2144dcb29eee47cead?d=http%3A%2F%2Fwww.gravatar.com%2Favatar%2F4e8e2db385219e064e6dea8fbd386434" alt="MSTROUT" style="display: block; margin: 0 3px 5px 0!important; border: 1px solid #666; border-radius: 3px; "> <span style="color: #333; font-weight: bold;">MSTROUT</span> </a> </span>
</div>

<div>
    </div></div>
</div>

# LICENSE

You can use and distribute this module under the same terms as Perl itself.
See the `LICENSE` file included in this distribution for complete
details.


Changes
Revision history for Mew

1.001005  2015-12-06
  - Use different prereq versions; old ones seem to cause test failures.

1.001004  2015-12-04
  - Fix missing prereqs (#3)
  - Do not use private Moo classes (#1)

1.001003  2015-11-30
  - Re-upload to fix indexing issue

1.001002  2015-11-27
  - Use saner code (mst)
  - Allow for chained attributes (mst)
  - Doc clarifications

1.001001  2015-11-27
  - First version released on an unsuspecting world




t/Class1.pm
package t::Class1;

use Mew;

has  _num     => PositiveNum;
has -_bool    => Bool;
has  _type    => ( Str, default => 'text/html');
has -_init    => ( Str, init_arg => 'initizer');
has -chained  => ( Str, chained => 1 );
has -chained2 => ( Num, chained => 1 );

has [qw/ar1 ar2/] => Str, ( required => 0, is => 'rw' );
# has [qw/_ar2/] => Str, ( required => 0 );

has  _cust => ( is => 'rw', default => 'Zoffix' );
1;

lib/ew.pm
package ew;

use strictures 2;
use Module::Runtime;

# VERSION

sub _mew {
  print <<'EOMEW';
                 ________________
                |                |_____    __
                |  I Love You!   |     |__|  |_________
                |________________|     |::|  |        /
   /\**/\       |                \.____|::|__|      <
  ( o_o  )_     |                      \::/  \._______\
   (u--u   \_)  |
    (||___   )==\
  ,dP"/b/=( /P"/b\
  |8 || 8\=== || 8
  `b,  ,P  `b,  ,P
    """`     """`

EOMEW
  exit 0;
}

BEGIN {
    my $package;
    sub import {
        _mew() if $0 eq '-';
        $package = $_[1] || 'Class';
        $package =~ s/^\+// and require_module $package;
    }
    use Filter::Simple sub { s/^/package $package;\nuse Mew;\n/; }
}

1;
__END__

=encoding utf8

=for stopwords oneliners

=head1 NAME

ew - syntactic sugar for Mew oneliners

=head1 SYNOPSIS

  perl -Mew=Foo -e 'has bar => Str, default => "baz"; print Foo->new->bar'

  # loads an existing class and re-"opens" the package definition
  perl -Mew=+My::Class -e 'print __PACKAGE__->new->bar'

=head1 DESCRIPTION

ew.pm is a simple source filter that adds C<package $name; use Mew;> to the
beginning of your script, intended for use on the command line via the -M
option.

=head1 SUPPORT

See L<Mew> for support and contact information.

=head1 AUTHORS

See L<Mew> for authors.

=head1 COPYRIGHT AND LICENSE

See L<Mew> for the copyright and license.

=cut


lib/Mew.pm
package Mew;

use strictures 2;

# VERSION

use Import::Into;
use Moo;

sub import {
    my $class = shift;
    strictures->import::into(1);
    Moo->import::into(1);
    MooX->import::into(1, $_) for @_;
    MooX::ChainedAttributes->import::into(1);
    Types::Standard->import::into(1, qw/:all/);
    Types::Common::Numeric->import::into(1, qw/:all/);

    my $target = caller;
    my $moo_has = $target->can('has');
    Moo::_install_tracked $target => has => sub {
        my $name_proto = shift;
        my @name_proto = ref $name_proto eq 'ARRAY'
            ? @$name_proto : $name_proto;

        my $mew_type;
        $mew_type = shift if @_ % 2 != 0;
        for my $attr ( @name_proto ) {
            my %spec = @_;
            if ( $mew_type ) {
                my %mew_spec;
                $mew_spec{required} = 1 unless $attr =~ s/^-//;
                ( $mew_spec{init_arg} = $attr ) =~ s/^_//
                    unless exists $spec{init_arg};

                %spec = (
                    is  => $spec{chained} ? 'rw' : 'ro' ,
                    isa => $mew_type,
                    %mew_spec,
                    %spec,
                );
            }
            $moo_has->( $attr => %spec );
        }

        return;
    };

    namespace::clean->import::into(1);
}

q|
    To err is human -- and to blame it on a computer is even more so
|;

__END__

=encoding utf8

=for stopwords Znet Zoffix Altreus copypasta mst

=head1 NAME

Mew - Moo with sugar on top

=head1 SYNOPSIS

=for pod_spiffy start code section

    #
    # This:
    #
    use Mew;

    has  _foo  => PositiveNum;
    has -_bar  => Bool;  # note the minus: it means attribute is not `required`
    has  type  => ( Str, default => 'html', chained => 1); # fluent interface
    has  _cust => ( is => 'ro', isa => sub{ 42 } );        # standard Moo `has`
    has [qw/_ar1  -_ar2/] => Str;                          # Multiple args

    #
    # Is the same as:
    #
    use strictures 2;
    use Types::Standard qw/:all/;
    use Types::Common::Numeric qw/:all/;
    use Moo;
    use MooX::ChainedAttributes;
    use namespace::clean;

    has _foo  => (
        init_arg => 'foo',
        is       => 'ro'
        isa      => PositiveNum,
        required => 1,
    );

    has _bar  => (
        init_arg => 'bar',
        is       => 'ro'
        isa      => Bool,
    );

    has type  => (
        chained  => 1,
        is       => 'rw'
        isa      => Str,
        default  => 'html',
    );

    has _cust => (
        is  => 'ro',
        isa => sub{ 42 },
    );

    has _ar1  => (
        init_arg => 'ar1',
        is       => 'ro'
        isa      => Str,
        required => 1,
    );

    has ar2  => (
        init_arg => 'ar2',
        is       => 'ro'
        isa      => Str,
    );


=for pod_spiffy end code section

=head1 DESCRIPTION

This module is just like regular L<Moo>, except it also imports
L<strictures> and L<namespace::clean>, along with
a couple of standard types modules. In addition, it sweetens the
L<Moo's has subroutine|Moo/has> to allow for more concise attribute
declarations.

=head1 READ FIRST

Virtually all of the functionality is described in L<Moo>.

=head1 IMPORTED MODULES

    use Mew;

Automatically imports the following modules: L<Moo>, L<strictures>,
L<Types::Standard>, L<Types::Common::Numeric>, L<MooX::ChainedAttributes>,
and L<namespace::clean>. B<NOTE: in particular the last one.> It'll scrub
your namespace, thus if you're using things like L<experimental>, you should
declare them B<after> you C<use Mew>.

=head1 C<has> SUGAR

=head2 Call it like if it were Moo

    has _cust => ( is => 'ro' );

First, you can call C<has> just like you'd call L<Moo/has> and it'll work
exactly as it used to. The sugar won't be enabled in that case.

=head2 Specify C<isa> type to get sugar

    has _cust => Str;
    has _cust => ( Str, default => "foo" ); # Note: can't use "=>" after Str
    has [qw/_z1  -z2/] => Str;

To get the sugar, you need to specify one of the imported types from either
L<Types::Standard> or L<Types::Common::Numeric> as the second argument. Once
that is done, C<Mew> will add some default settings, which are:

    1) Set `isa` to the type you gave
    2) Set `is` to 'ro' (or 'rw', if `chained` is set)
    3) Set `require` to 1
    4) Set `init_arg` to the name of the attribute, removing
        the leading underscore, if it's present

Thus, C<< has _cust => Str; >> is equivalent to

    use Types::Standard qw/Str/;
    has _cust => (
        init_arg => 'cust',
        is       => 'ro'
        isa      => Str,
        required => 1,
    );

You can specify same settings for multiple attributes by providing
their names in an arrayref:

    has [qw/_z1  -z2/] => Str;

=for pod_spiffy start warning section

B<IMPORTANT NOTE:> because Perl's fat comma (C<< => >>) quotes the argument
on the left side, using it after the type won't work:

    has _cust => Str => ( default => "BROKEN" ); # WRONG!!!!
    has _cust => Str, ( default => "WORKS" ); # Correct!
    has _cust => ( Str, default => "WORKS" ); # This is fine too

=for pod_spiffy end warning section

=head2 Method chaining

    package Foo;
    use Mew;
    has cost   => ( PostiveNum, chained => 1 );
    has weight => ( PostiveNum, chained => 1 );
    has size   => ( Str,        chained => 1 );

    ...

    my $object = Foo->new->cost( 42 )->weight( 45 )->size("X-Large");
    say $object->size; # prints "X-Large"

To have L<fluent interface|https://en.wikipedia.org/wiki/Fluent_interface>
or allow "chaining" your attributes, simply add C<< chained => 1 >> option
to your attribute declaration. B<Note:> this will automatically use
C<rw> instead of C<ro> for the default of the C<is> option.

=head3 Modify the sugar

It's possible to alter the defaults created by C<Mew>:

=head4 Remove C<required>

    has -_cust => Str;

Simply prefix the attribute's name with a minus sign to avoid setting
C<< required => 1 >>.

=head4 Modify other options

    has  _cust => Str, ( init arg => "bar" );
    has -_cust => Str, ( is => "lazy" );

You can explicitly provide values for options set by C<Mew>, in which case
the values you provide will be used instead of the defaults.

=head1 SEE ALSO

L<Moo>, L<Type::Tiny>

=for pod_spiffy hr

=head1 REPOSITORY

=for pod_spiffy start github section

Fork this module on GitHub:
L<https://github.com/zoffixznet/Mew>

=for pod_spiffy end github section

=head1 BUGS

=for pod_spiffy start bugs section

To report bugs or request features, please use
L<https://github.com/zoffixznet/Mew/issues>

If you can't access GitHub, you can email your request
to C<bug-Mew at rt.cpan.org>

=for pod_spiffy end bugs section

=head1 AUTHOR

Part of the code was borrowed from L<Moo>'s innards. L<ew> module is an
almost-verbatim copy of L<oo> module. Thanks to I<Matt S. Trout (mst)> for
changing my copypasta of Moo's internals to sane code and other help.
Props to I<Altreus> for coming up with the name for the module.

The rest is:

=for pod_spiffy start author section

=for pod_spiffy author ZOFFIX

=for pod_spiffy end author section

=head1 CONTRIBUTORS

=for pod_spiffy start contributors section

=for pod_spiffy author MSTROUT

=for pod_spiffy end contributors section

=head1 LICENSE

You can use and distribute this module under the same terms as Perl itself.
See the C<LICENSE> file included in this distribution for complete
details.

=cut